pkg/core/chain/chain.go

```diff
@@ -176,115 +152,138 @@ func New(ctx context.Context, eventBus *eventbus.EventBus, rpcBus *rpcbus.RPCBus
-			if err := c.onAcceptBlock(m); err != nil {
-				log.WithError(err).WithField("topic", topics.Block.String()).Warn("Handling block failed")
-			}
-		case m := <-c.initializationChan:
-			if err := c.onInitialization(m); err != nil {
-				log.WithError(err).WithField("topic", topics.Initialization.String()).Warn("Handling initialization failed")
-			}
-		case certificateMsg := <-c.certificateChan:
-			c.handleCertificateMessage(certificateMsg)
-		case height := <-c.highestSeenChan:
-			// TODO: check out if highestSeen could be out of chain
-			c.highestSeen = height
-		case r := <-c.verifyCandidateBlockChan:
-			c.processCandidateVerificationRequest(r)
-		case r := <-c.getLastCertificateChan:
-			c.provideLastCertificate(r)
-		case r := <-c.getLastCommitteeChan:
-			c.provideLastCommittee(r)
-		case <-c.ctx.Done():
-			// TODO: dispose the Chain
-		}
-	}
+// SetupConsensus adds the missing fields on the Chain which need to be populated
+// by the user. Once the fields are populated, consensus is started.
+func (c *Chain) SetupConsensus(pk keys.PublicKey, blsKeys key.Keys) error {
+	c.lock.Lock()
+	c.pubKey = &pk
+	e := &consensus.Emitter{
+		EventBus:    c.eventBus,
+		RPCBus:      c.rpcBus,
+		Keys:        blsKeys,
+		Proxy:       c.proxy,
+		TimerLength: config.ConsensusTimeOut,
+	}
+
+	c.loop = loop.New(e)
+	c.lock.Unlock()
+	return c.startConsensus()
 }
 
-func (c *Chain) beginAccepting(blk *block.Block) bool {
-
-	field := logger.Fields{"process": "onAcceptBlock", "height": blk.Header.Height}
-	lg := log.WithFields(field)
+// ProcessBlock will handle blocks incoming from the network. It will allow
+// the chain to enter sync mode if it detects that we are behind, which will
+// cancel the running consensus loop and attempt to reach the new chain tip.
+// Satisfied the peer.ProcessorFunc interface.
+func (c *Chain) ProcessBlock(m message.Message) ([]*bytes.Buffer, error) {
+	blk := m.Payload().(block.Block)
+	log.WithField("height", blk.Header.Height).Trace("received block")
 
-	// Ignore blocks from peers if we are only one behind - we are most
-	// likely just about to finalize consensus.
-	// TODO: we should probably just accept it if consensus was not
-	// started yet
+	c.lock.Lock()
+	// Is it worth looking at this?
+	if blk.Header.Height <= c.tip.Header.Height {
+		log.Debug("discarded block from the past")
+		c.lock.Unlock()
+		return nil, nil
+	}
 
-	if !c.counter.IsSyncing() {
-		lg.Warn("could not accept block since we are syncing")
-		return false
+	if blk.Header.Height > c.highestSeen {
+		c.highestSeen = blk.Header.Height
 	}
 
 	// If we are more than one block behind, stop the consensus
-	lg.Debug("topics.StopConsensus")
+	log.Debug("topics.StopConsensus")
 	// FIXME: this call should be blocking
-	errList := c.eventBus.Publish(topics.StopConsensus, message.New(topics.StopConsensus, message.EMPTY))
-	diagnostics.LogPublishErrors("chain/chain.go, topics.StopConsensus", errList)
+	if c.cancel != nil {
```

Comment:
Is it true that any node who sends a forged block with higher block height is capable of cancel/disabling consensus loop for awhile? 

Without a sanity or certificate check for the wire (block) message, a malicious  node might be capable to suspend consensus for majority of the network for the benefit of his/her own provisioners/generators.  

NB I raised this on a few stand-ups but not sure if it is ignored/forgotten.

Reply:
That's right, but I don't think this is an issue with the architecture per se. With the old `Synchronizer` approach, the same problem occurs. Perhaps this is something we should extract into another issue?

Reply:
It's an architecture thing but we always can postpone it.

Reply:
Agreed on postponing, but the issue must be recorded. @ggoranov would you mind creating an issue for this?